### ISSUE TITLE
Can't make it work with an ajax upload

### ISSUE THREAD

**Issue by vikingmute** [expressjs/multer#193]
I am trying to do the image ajax upload using Express4 and multer, but it doesn't work

here is my request screenshot
![image](https://cloud.githubusercontent.com/assets/343954/9128116/9c44f2bc-3cf6-11e5-8b58-7330a9aea097.png)

my client code 
```javascript
import Ember from 'ember';

export default Ember.TextField.extend({
  type: 'file',
  change: function(e) {
    var inputFiles = e.target.files;
    var inputFile = inputFiles[0];
    var formData = new FormData();
    formData.append('photo', inputFile);
    Ember.$.ajax({
      type: 'POST',
      url: '/upload',
      data: formData,
      cache: false,
      contentType: false,
      processData: false,
      success: function(data) {
        console.log(data);
      },
      error: function(err) {
        console.error(err)
      }
    })
  }
});
```

and server code
```javascript
var multer  = require('multer');
var upload = multer({ dest: './uploads/' });
var handler = upload.single('photo');
//split upload endpoint out
server.post('/upload', handler, function (req, res, next) {
  // req.file is the `avatar` file
  console.log(req.file);
});
```

Thx for your help!

**Comment by LinusU** [comment]
What is the problem you are seeing? Do you get an error? Is `req.file` undefined? Or is `req.file` an empty file?

From the request screenshot it seems like it isn't sending any data, aka. an empty file.

**Comment by vikingmute** [comment]
@LinusU there is no error, req.file is undefined,  
Is there anything wrong with my client code? the screenshot shows it is sending a file named 'photo' and filename is 'img_1439.jpg'

**Comment by LinusU** [comment]
That is true, but it shows the boundary right after, no data that is the actual image.

Could you try making a simple form and see if that works for you? Then we know where the problem is.

```html
<form action="/upload" enctype="multipart/form-data">
  <input type="file" name="photo" />
  <input type="submit" />
</form>
```

**Comment by kishanj918** [comment]
I am also getting the same while submitting form through ajax.. I am getting the following error.
Unhandled rejection TypeError: Cannot read property 'filename' of undefined

**Comment by vikingmute** [comment]
@LinusU I have test the simple form demo, it works fine! I think there might be a bug when using ajax to upload.

**Comment by LinusU** [comment]
@kishanj918 Do you have a stack trace for that? Where do you see the error?

@vikingmute Could you try it with native ajax then instead of the ember wrapped part?

```javascript
var req = new XMLHttpRequest()
var file = new Blob(['This is a test'], { type: 'text/plain' })
var data = new FormData()

data.append('photo', file, 'test.txt')

req.open('POST', '/upload')
req.send(data)
```

**Comment by vikingmute** [comment]
@LinusU  hey, I have used the native ajax to test it, it works fine.
network shot:
![image](https://cloud.githubusercontent.com/assets/343954/9188804/379eb4c2-4013-11e5-86ef-a094c869ce90.png)

Ember's ajax is based on jQuery, the only different should be the Content-Type header, do you think this will cause the issue?

**Comment by LinusU** [comment]
Yep, jQuery is at fault here.

- [Workaround at Stack Overflow](http://stackoverflow.com/questions/6974684/how-to-send-formdata-objects-with-ajax-requests-in-jquery)
- [Bugreport at jQuery](http://bugs.jquery.com/ticket/9995)

**tl;dr** Add `processData: false` and `contentType: false` to jQuery AJAX options.

**Comment by vikingmute** [comment]
hey, @LinusU  
I do add processData:false and contentType:false in the jQuery ajax options at the first time.  But it won't work.

Anyway, I am using the raw XMLHttpRequest to do it, everything works fine for now.

I truly appreciate your help in solving this problem.

**Comment by LinusU** [comment]
Seems smart, [it should work in IE >= 10 and all other browsers](http://caniuse.com/#feat=xhr2)

**Comment by dossan** [comment]
I have a similar problem. I am trying to send a file using XMLHttpRequest, but req.file is always `undefined`. I have asked the following question on SO:

> http://stackoverflow.com/questions/34091195/uploading-a-file-with-formdata-and-multer

You can see there more details. I still didn't manage to make it work.

**Comment by conan19842002** [comment]
Yes, I have the same issue: req.file always returns undefined. I really don't know what went wrong!
